### ISSUE TITLE
Make `SyntaxHighlighter.highlight!` to return partially colorized string in case of error.

### ISSUE THREAD

**Issue by I3oris** [crystal-lang/crystal#12689]
## Feature Request

Currently, the crystal highlighter (`Crystal::SyntaxHighlighter::Colorize`) present two functions, `highlight` that colorizes a code or raises on error, or `highlight!` that colorize or returns the original uncolored code. This former behavior is not very useful if used in a REPL, because it will uncolorizes all the code in case of partial string. See below:

![uncolorized_string](https://user-images.githubusercontent.com/37157434/198843902-1cf3ca4a-acae-4f43-8a8e-008abd1694cf.gif)

I propose to modify the function in order to return a partially colorized string in case of error, which in this case lead to better results:
![colorized_string](https://user-images.githubusercontent.com/37157434/198844229-9b5cf9c6-fde9-4a1d-b203-2e7250120009.gif)

I see three possibilities:
1) Modify `highlight!` for that return partially colorized string on error, and if anyone want the old behavior it could do it simply with `highlight(code) rescue code`. However it's a breaking change.
2) Add the feature in a new function like `highlight_or_partial!` (naming could be better).
3) Doing nothing, but the shard like [reply](https://github.com/I3oris/reply) have to monkey patch the highlighter to fix the behavior.
 
For consistency, I propose to apply the change also for (`Crystal::SyntaxHighlighter::HTML`).

I also prepared a PR for this, I will just wait which possibility we decide it's the best.

**Comment by straight-shoota** [comment]
I'm not sure if this needs a change in `SyntaxHighlighter` A full implementation of partial syntax highlighting is very complex and not really necessary for this use case.
Instead, I think it would be a better option to handle this directly in the REPL. If highlighting the full code fails, it can just re-use the previously successful highlighting for the beginning substring.

**Comment by I3oris** [comment]
If for 'partially' we actually mean 'colorized until the error, then uncolorized after'
the implementation is not very complex.
We only have to add a member `@pos` on the lexer and incrementing it every time the lexer goes forward.
Then in case of error we can just append the part after `@pos` to the already colorized string. 
As follow: 

```crystal
Crystal::SyntaxHighlighter
  @pos = 0

  def highlight
     ...
    @pos = lexer.current_pos
    token = lexer.next_token
    ...
  end
end

class Crystal::SyntaxHighlighter::Colorize < Crystal::SyntaxHighlighter
  def self.highlight!(code : String)
    String.build do |io|
      new(io).highlight!(code)
    end
  end

  def highlight!(code : String)
    highlight(code)
  rescue
    @io << String.new(code.to_slice[@pos..])
  end
end
```


However I hadn't consider your idea of 'just re-use the previously successful' but this could be fine too, thank!

**Comment by straight-shoota** [comment]
This definition of partial fits well for the interpreter's REPL, but for a general implementation it wouldn't be very great. Think of use cases such as an IDE or code examples on a website.
So I think it's better to implement an improvement specific to the REPL and do that directly at the place where it's used.

**Comment by asterite** [comment]
In an IDE you might also have partial input. I think there's no point in not doing this if it's easy.

**Comment by straight-shoota** [comment]
I would generally expect from a syntax highlighter to highlight all the code that can be highlighted, and not break at the first syntax error, leaving everything behind without highlight.

Example:
```cr
class Foo
  def initialize
    @
  end
end
```

Note that the `end` keywords are highlighted. With the proposed implementation of partial highlighting, they wouldn't because the parser breaks at `@`.

**Comment by asterite** [comment]
> With the proposed implementation of partial highlighting, they wouldn't because the parser breaks at @.

Aside from the proposed implementation, do you think it's a good idea to support this? I've been playing around with REPLY and having the syntax be colored even if there are errors would greatly improve the experience.

**Comment by straight-shoota** [comment]
Yes, certainly! It's definitely a good and very useful improvement, even if it doesn't cover partial edge cases. All I'm saying since https://github.com/crystal-lang/crystal/issues/12689#issuecomment-1295913489 is just that I'd prefer to implement the logic for partial highlighting in the interpreter's REPL, and not make it a generic feature of `SyntaxHighlighter`.
